planner/core/rule_constant_propagation.go

```diff
@@ -0,0 +1,284 @@
+	}
+	result := make([]expression.Expression, 0, len(candidateConstantPredicates))
+	for _, predicate := range candidateConstantPredicates {
+		// The column of predicate must exist in projection exprs
+		columns := expression.ExtractColumns(predicate)
+		// The number of columns in candidate predicate must be 1.
+		if len(columns) != 1 {
+			continue
+		}
+		if replace[string(columns[0].HashCode(nil))] == nil {
+			// The column of predicate will not appear on the upper level
+			// This means that this predicate does not apply to the constant propagation optimization rule
+			// For example: select * from t, (select b from s where s.a=1) tmp where t.b=s.b
+			continue
+		}
+		clonePredicate := predicate.Clone()
+		ResolveExprAndReplace(clonePredicate, replace)
+		result = append(result, clonePredicate)
+	}
+	return result
+}
+
+// validComparePredicate checks if the predicate is an expression like [column '>'|'>='|'<'|'<='|'=' constant].
+// return param1: return true, if the predicate is a compare constant predicate.
+// return param2: return the column side of predicate.
+func validCompareConstantPredicate(candidatePredicate expression.Expression) bool {
+	scalarFunction, ok := candidatePredicate.(*expression.ScalarFunction)
+	if !ok {
+		return false
+	}
+	if scalarFunction.FuncName.L != ast.GT && scalarFunction.FuncName.L != ast.GE &&
+		scalarFunction.FuncName.L != ast.LT && scalarFunction.FuncName.L != ast.LE &&
+		scalarFunction.FuncName.L != ast.EQ {
+		return false
+	}
+	column, _ := expression.ValidCompareConstantPredicateHelper(scalarFunction, true)
+	if column == nil {
+		column, _ = expression.ValidCompareConstantPredicateHelper(scalarFunction, false)
+	}
+	if column == nil {
+		return false
+	}
+	return true
+}
+
+// Add a new selection between parent plan and current plan with candidate predicates
+/*
++-------------+                                    +-------------+
+| parentPlan  |                                    | parentPlan  |
++-----^-------+                                    +-----^-------+
+      |           --addCandidateSelection--->            |
++-----+-------+                              +-----------+--------------+
+| currentPlan |                              |        selection         |
++-------------+                              |   candidate predicate    |
+                                             +-----------^--------------+
+                                                         |
+                                                         |
+                                                    +----+--------+
+                                                    | currentPlan |
+                                                    +-------------+
+*/
+// If the currentPlan at the top of query plan, return new root plan (selection)
+// Else return nil
+func addCandidateSelection(currentPlan LogicalPlan, currentChildIdx int, parentPlan LogicalPlan,
```

Comment:
And then to let the `rule_predicate_push_down` push these new predicted down through the `join`?

Reply:
Yes